crates/router/src/core/payment_methods/cards.rs

```diff
@@ -1449,10 +1449,11 @@ pub async fn vault_payment_method_in_locker(
     state: &routes::SessionState,
     merchant_account: &domain::MerchantAccount,
     pmd: &api::PaymentMethodCreateData,
+    existing_vault_id: Option<String>,
 ) -> errors::RouterResult<pm_types::AddVaultResponse> {
     let payload = pm_types::AddVaultRequest {
         entity_id: merchant_account.get_id().to_owned(),
-        vault_id: uuid::Uuid::now_v7().to_string(),
+        vault_id: existing_vault_id.unwrap_or(uuid::Uuid::now_v7().to_string()),
         data: pmd.clone(),
         ttl: state.conf.locker.ttl_for_storage_in_secs,
     }
```

Comment:
Don't handle everthing as Internal Server Error, `vault_payment_method_in_locker` function need not return `ApiErrorResponse` instead it should return `VaultingError`

VaultingError will consists of `EncodingFaile`,`VaultingFailed`, `ResponseParsingError`

Reply:
In the `retrieve_payment_method_from_vault` do the repespective error related changes